blockingconcurrentqueue.h

```diff
@@ -452,19 +476,23 @@ class BlockingConcurrentQueue
 	// includes making the memory effects of construction visible, possibly with a
 	// memory barrier).
 	explicit BlockingConcurrentQueue(size_t capacity = 6 * BLOCK_SIZE)
-		: inner(capacity), sema(create<LightweightSemaphore>(), &BlockingConcurrentQueue::template destroy<LightweightSemaphore>)
+		: inner(capacity),
+      sema_dequeue(create<LightweightSemaphore>(), &BlockingConcurrentQueue::template destroy<LightweightSemaphore>),
+      sema_enqueue(create<LightweightSemaphore>(), &BlockingConcurrentQueue::template destroy<LightweightSemaphore>)
```

Comment:
`sema_enqueue` starts off with a size `0` capacity; it should start with the number of free slots instead. However, since the actual number of free slots varies depending on `BLOCK_SIZE` and whether an implicit or explicit producer is being queued to, it's actually impossible to know whether an `enqueue` will succeed or not unless the maximum number of concurrent implicit & explicit producers are known (to put a bound on the available number of slots at any given time). Even then, there's usually several wasted blocks. I'm not sure how this would work...